### PR TITLE
Feature/user crud operations

### PR DESCRIPTION
**Pull request by Deepak-webelight**
- post `/:id` get user data
- patch `/:id` update user data 
- delete `/:id` delete user data

### CODE REVIEW

**Comment by Deepak-webelight** [comment]
**Note** after merge `feature/rbac` branch required to use `guards` to protects above routes

File: server/src/modules/user/user.controller.ts [hunk 1]
```diff
@@ -106,31 +118,62 @@ export class UserController {
...
+  ): Promise<IUserResponse> {
+    try {
+      // call user service to extract user information
+      const user = await this.userService.getUserById(id);
+
+      return {
+        message: 'User information fetched successfully',
+        data: user,
+        status: HttpStatus.OK,
+      };
```
Comment (anshs-webelight) [inline server/src/modules/user/user.controller.ts]: You can create the common response-handler interceptor and send common response syntax for all the endpoints

File: server/src/modules/user/user.controller.ts [hunk 2]
```diff
@@ -106,31 +118,62 @@ export class UserController {
...
+    } catch (err) {
+      throw new BadRequestException(err.response);
+    }
+  }
+  @Patch(':id')
```
Comment (anshs-webelight) [inline server/src/modules/user/user.controller.ts]: leave one line beyween two endpoints

File: server/src/modules/user/user.controller.ts [hunk 3]
```diff
@@ -106,31 +118,62 @@ export class UserController {
...
+  async updateUserDetails(
+    @Param() { id }: getUserDetailsByIdDto,
+    @Body() body: UpdateUserDataRequestBodyDto,
+  ): Promise<IUserResponse> {
+    try {
+      // call user service to find and update user id
+      const updatedUser = await this.userService.updateUserDetails(id, body);
```
Comment (anshs-webelight) [inline server/src/modules/user/user.controller.ts]:
```suggestion
      // call user service to find and update user id
      return this.userService.updateUserDetails(id, body);
```

File: server/src/modules/user/user.controller.ts [hunk 4]
```diff
@@ -106,31 +118,62 @@ export class UserController {
...
+      console.log(updatedUser, 'updatedUser');
+      return {
+        message: 'Updated user details',
+        data: updatedUser,
+      };
```
Comment (anshs-webelight) [inline server/src/modules/user/user.controller.ts]: move this return statement to service file

Comment (anshs-webelight) [reply server/src/modules/user/user.controller.ts]: Do this for all the endpoints

File: server/src/modules/user/user.controller.ts [hunk 5]
```diff
@@ -106,31 +118,62 @@ export class UserController {
...
 
-  @PublicRoute()
-  @SuperAdmin()
-  @Post('/admin')
-  async createNewAdminUser(
-    @Body() body: SignUpRequestBodyDto,
+  // get user information by id
+  @Get(':id')
+  async getUserById(
+    @Param() { id }: getUserDetailsByIdDto,
```
Comment (anshs-webelight) [inline server/src/modules/user/user.controller.ts]:
```suggestion
    @Param("id") id: number,
```

Comment (anshs-webelight) [reply server/src/modules/user/user.controller.ts]: No need to create `getUserDetailsByIdDto`

File: server/src/modules/user/user.service.ts [hunk 6]
```diff
@@ -3,9 +3,14 @@ import { JwtService } from '@nestjs/jwt';
...
-import { LoginRequestDto, SignUpRequestBodyDto } from './user.dto';
+import {
+  LoginRequestDto,
+  SignUpRequestBodyDto,
+  UpdateUserDataRequestBodyDto,
+} from './user.dto';
 import { User } from './user.model';
 import { createHashPassword, verifyPassword } from 'src/utils/bycrpt';
+import { IUpdateUserDetailsfilter } from './user.interface';
 import { Role } from '../../guards/role-auth.guard';
```
Comment (anshs-webelight) [inline server/src/modules/user/user.service.ts:18]: use absolute path for imports

File: server/src/modules/user/user.service.ts [hunk 7]
```diff
@@ -102,6 +107,41 @@ export class UserService {
     }
   }
 
+  async getUserById(id: string): Promise<User> {
+    try {
+      return await this.userModel.findById(id, {
```
Comment (anshs-webelight) [inline server/src/modules/user/user.service.ts]:
```suggestion
      return this.userModel.findById(id, {
```
await with return statement not needed

File: server/src/modules/user/user.service.ts [hunk 8]
```diff
@@ -102,6 +107,41 @@ export class UserService {
...
+        role: 0,
+        password: 0,
+      });
+    } catch (err) {
+      throw new BadRequestException('User not found');
+    }
+  }
+
+  async updateUserDetails(id: string, body: UpdateUserDataRequestBodyDto) {
+    try {
```
Comment (anshs-webelight) [inline server/src/modules/user/user.service.ts]: why are we adding a try-catch block in the service file? adding this comment for the second time!

File: server/src/app.controller.ts [hunk 9]
```diff
@@ -1,12 +1,10 @@
...
 import { AppService } from './app.service';
-import { PublicRoute } from './guards/auth.guard';
-import {appConfig}  from './appConfig/configuration';
-
+import { AuthGuard } from './guards/auth.guard';
 @Controller()
 export class AppController {
   constructor(private readonly appService: AppService) {}
-  @PublicRoute()
+  @UseGuards(AuthGuard)
```
Comment (anshs-webelight) [inline server/src/app.controller.ts]: why do we need AuthGuard here?

File: server/src/app.module.ts [hunk 10]
```diff
@@ -31,8 +34,12 @@ import { SuperAdminAuthGuard } from './guards/superAdmin.auth.guard';
       useClass: RolesGuard,
```
Comment (anshs-webelight) [inline server/src/app.module.ts:29]: import of Guard not needed

Comment (anshs-webelight) [reply server/src/app.module.ts:29]: can you please try without importing guards in app.module.ts file?

File: server/src/intercepter/response.intercepter.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+import { Observable, throwError } from 'rxjs';
+
+// Response by routes 
+@Injectable()
+export class ResponseInterceptor<T> implements NestInterceptor<T, { data: T }> {
+  intercept(
+    context: ExecutionContext,
+    next: CallHandler,
+  ): Observable<{ data: T }> {
+    return next.handle().pipe(map((data) => ({ data })));
```
Comment (anshs-webelight) [inline server/src/intercepter/response.intercepter.ts:20]: is there any impact in response after using response Interceptor and without interceptor?

File: server/src/modules/user/user.controller.ts [hunk 12]
```diff
@@ -1,136 +1,138 @@
...
+      return this.userService.logout(res);
     } catch (error) {
       throw new BadRequestException(error.response);
     }
   }
 
+  // refresh an expired token
   @Post('refresh')
-  async refreshToken(
+  refreshToken(
```
Comment (anshs-webelight) [inline server/src/modules/user/user.controller.ts]: why we remove async from this endpoint?

File: server/src/modules/user/user.controller.ts [hunk 13]
```diff
@@ -1,136 +1,138 @@
...
-      res.cookie('token', tokenFormat(token), cookieConfiguration);
-      res.cookie(
-        'refreshToken',
-        tokenFormat(refreshToken),
-        cookieConfiguration,
-      );
+  // update user details
+  @Patch()
+  async updateUserDetails(
+    @Query('id') id: string,
```
Comment (anshs-webelight) [inline server/src/modules/user/user.controller.ts]: should come from params or from token directly

File: server/src/modules/user/user.controller.ts [hunk 14]
```diff
@@ -1,136 +1,138 @@
...
 
-  @PublicRoute()
-  @SuperAdmin()
-  @Post('/admin')
-  async createNewAdminUser(
-    @Body() body: SignUpRequestBodyDto,
-    @Res({ passthrough: true }) res: Response,
-  ): Promise<IUserResponse> {
+  // get user information by id
+  @Get()
```
Comment (anshs-webelight) [inline server/src/modules/user/user.controller.ts:99]: from token or params

File: server/src/modules/user/user.controller.ts (hunk 13 above)
Comment (anshs-webelight) [reply server/src/modules/user/user.controller.ts]: we should add auth guard here and take userId from token

File: server/src/modules/user/user.controller.ts [hunk 15]
```diff
@@ -1,136 +1,138 @@
...
-      return {
-        message: 'Admin User Created Successfully',
-        status: HttpStatus.CREATED,
-      };
-    } catch (error) {
-      throw new BadRequestException(error.response);
+  // delete user
+  @Delete()
+  async deleteUser(
+    @Query('id') id: string,
```
Comment (anshs-webelight) [inline server/src/modules/user/user.controller.ts]: we should add auth guard here and take userId from token

File: server/src/modules/user/user.service.ts [hunk 16]
```diff
@@ -63,70 +83,169 @@ export class UserService {
...
+    const { res, refreshToken, token } = data;
+    if (token) res.cookie('token', tokenFormat(token), cookieConfiguration);
+    if (refreshToken)
+      res.cookie(
+        'refreshToken',
+        tokenFormat(refreshToken),
+        cookieConfiguration,
+      );
+  }
+  logout(res: Response) {
```
Comment (anshs-webelight) [inline server/src/modules/user/user.service.ts]: leave blank line between two service's method

**Review by anshs-webelight — CHANGES_REQUESTED** [review]

File: server/src/app.controller.ts (hunk 9 above)
Comment (Deepak-webelight) [reply server/src/app.controller.ts]: removed authGuard implemented @PublicRoute() decorater

File: server/src/app.module.ts (hunk 10 above)
Comment (Deepak-webelight) [reply server/src/app.module.ts:29]: implemented applied guards manually removed from globally

File: server/src/intercepter/response.intercepter.ts (hunk 11 above)
Comment (Deepak-webelight) [reply server/src/intercepter/response.intercepter.ts:20]: yes while using this our all response data is getting in 

```
{
  data : {
    // response 
  }

}
```

and in error one it responde error : { }

**Pull request merged**